src/Check/DoctrineMigration.php

```diff
@@ -17,70 +17,96 @@
-            $this->migratedVersions = $this->getMigratedVersionsFromDependencyFactory($input);
-            return;
+        if (! $input instanceof DependencyFactory && ! $input instanceof Configuration) {
+            throw new InvalidArgumentException(<<<'MESSAGE'
+                Invalid Argument for DoctrineMigration check.
+                If you are using doctrine/migrations ^3.0,
+                pass Doctrine\Migrations\DependencyFactory as argument.
+                If you are using doctrine/migrations ^2.0,
+                pass Doctrine\Migrations\Configuration\Configuration as argument.
+                MESSAGE
+            );
         }
 
-        // check for doctrine/migrations:^2.0
-        if ($input instanceof Configuration
-            && method_exists($input, 'getAvailableVersions')
-            && method_exists($input, 'getMigratedVersions')
-        ) {
-            $this->availableVersions = $input->getAvailableVersions();
-            $this->migratedVersions = $input->getMigratedVersions();
-            return;
-        }
-
-        throw new InvalidArgumentException(<<<'MESSAGE'
-            Invalid Argument for DoctrineMigration check.
-            If you are using doctrine/migrations ^3.0, pass Doctrine\Migrations\DependencyFactory as argument.
-            If you are using doctrine/migrations ^2.0, pass Doctrine\Migrations\Configuration\Configuration as argument.
-            MESSAGE
-        );
+        $this->input = $input;
     }
 
     /**
      * Perform the actual check and return a ResultInterface
      */
     public function check(): ResultInterface
     {
-        $notMigratedVersions = array_diff($this->availableVersions, $this->migratedVersions);
+        $availableVersions = $this->getAvailableVersionsFromInput($this->input);
+        $migratedVersions = $this->getMigratedVersionsFromInput($this->input);
+
+        $notMigratedVersions = array_diff($availableVersions, $migratedVersions);
         if (! empty($notMigratedVersions)) {
             return new Failure('Not all migrations applied', $notMigratedVersions);
         }
 
-        $notAvailableVersion = array_diff($this->migratedVersions, $this->availableVersions);
+        $notAvailableVersion = array_diff($migratedVersions, $availableVersions);
         if (! empty($notAvailableVersion)) {
             return new Failure('Migrations applied which are not available', $notMigratedVersions);
         }
 
         return new Success();
     }
 
+    /**
+     * @param DependencyFactory|Configuration $input
+     * @return Version[]
+     */
+    private function getAvailableVersionsFromInput($input): array
+    {
+        if ($this->input instanceof DependencyFactory) {
+            return $this->getAvailableVersionsFromDependencyFactory($input);
+        }
+
+        if ($this->input instanceof Configuration) {
+            return $this->input->getAvailableVersions();
+        }
+
+        throw new LogicException(<<<'MESSAGE'
```

Comment:
Not sure if this is a good way to handle this. Please suggest any improvements for handling the theoretical case that, despite the constructor validation, $input is neither `DependencyFactory` not `Configuration`

Reply:
If the constructor parameter type is properly specified, no check is really needed

Reply:
Since `$this->input` is either a `DependencyFactory` or a `Configuration`, you can remove the second `if ()` completely, and just do `return $this->input->getAvailableVersions();` :+1: